### ISSUE TITLE
Clarify ContainerStatus.Started null value and initialize the value for Init containers

### ISSUE THREAD

**Issue by SergeyKanzhelev** [kubernetes/kubernetes#115553]
This issue is to discuss whether we can change the `Started` field of `ContainerStatus` from pointer to non-pointer bool value. The pointer type is inconsistent with other field `Ready`. The semantic of `Started` being `null` is undefined and it is not ideal if users will use `Started` field `null` value as indication of a spacial condition beyond what `false` indicates. So if there is a process of making field non-pointer after skew period has passed, it would be great to follow it.

The field was introduced as a pointer following [this suggestions](https://github.com/kubernetes/kubernetes/pull/77807#discussion_r289542539:): 

> If this is optional, it needs to be a pointer. We can't add non-optional fields, ergo this should be a pointer. Otherwise an old client will not set it and it will be false, which is super confusing.

The field long passed it's skew inconsistency period.

The field is set by prober_manager: https://github.com/kubernetes/kubernetes/blob/0e3818e02760afa8ed0bea74c6973f605ca4683c/pkg/kubelet/prober/prober_manager.go#L270

## Current behavior

The PR: https://github.com/kubernetes/kubernetes/pull/115463 improves the documentation for the `Started` field of `ContainerStatus`. We never declared when `null` value means for the `ContainerStatus`.

For regular container:
- Field is `null` before the `prober_manager` starts.
- Field is `false` once `prober_manager` started.
- Field is `true` when no startup probes defined or once startup probe succeeded.

For Init and ephemeral containers:
- Field is always `null` as those do not start the `prober_manager`.

`StartedAt` field of `Status` type is always populated for both - regular and Init containers. It is populated earlier than `Started` will be set by `prober_manager`. So there is period of time when `StartedAt` has a date and `Started` is `null`.

There is also `Ready` field on `ContainerStatus` that is not a pointer. It will be `false` from the Pod creation and will change to `true` while readiness probes succeed.

## Changes for sidecar containers

The question about this field came up because we plan to use it for https://github.com/kubernetes/enhancements/issues/753. Also resolving this would clarify the field that needs to be added as part of that same KEP.

We need a signal for sidecar containers KEP to indicate that the sidecar container started and other containers initialization can proceed. We believe that relying on `Started` field would be the best option. So we will initialize the `Started` field for sidecar containers and there will be logic relying on it.

## Proposal

- Change `Started` type from `*bool` to `bool`.
- This changes behavior for Init Containers and Pods before they received by the kubelet - value will be `false` instead of `null`
- Clearly document the difference between `ContainerStatus.Started` and `ContainerStatus.State.StartedAt` fields.
- For sidecar container KEP use the `Started` field as a signal that sidecar container started and ready to proceed with other containers initialization.

/sig architecture
/sig node
/label api-review
//cc @matthyx @dims 

Discussion in Slack: https://kubernetes.slack.com/archives/CJUQN3E4T/p1675467718491919

**Comment by dims** [comment]
Thanks @SergeyKanzhelev will defer to the API reviewers

**Comment by matthyx** [comment]
Thanks for opening this @SergeyKanzhelev !
I'm pretty sure it's my omission when I moved the startupProbe to GA.

@thockin do you agree this was supposed to happen when we worked on the KEP?

**Comment by thockin** [comment]
There are four ways to set default values.

1) Make it a pointer, set it from your controller.  This is not really a "default" because clients can observe the field as unspecified (nil).

2) Make it a pointer, set it from defaults.go.  Clients should never see the field as unspecified.  This is necessary when the Go "zero value" is in-bounds for the field (to distinguish zero-value from unspecified)

3) Don't make it a pointer, set it from defaults.go.  Clients should never see the field as unspecified.  This only works of the Go "zero-value" is out of bounds for the field.

4) Don't make it a pointer, let the Go zero-value be the default.

I do not see explicit default-setting code for this field, so it is either case (1) or case (4).  It is a pointer, so it must be case (1).  IIUC you are proposing to make it case (4).  I write this to call out that (2) is also available to you.

Considerations:
* Changing the field to a non-pointer is "simpler", especially if you never defined what it means to be unspecified.
* Changing the field to a non-pointer is a breaking change for client-go users, but not k8s API users.
* The fact that we have `StartedAt` and `Started` kind of makes me grumpy, but no fixing it now.

We usually argue for (2) because of consistency ("all optional fields are pointers, period") but it is not UNIVERSALLY upheld by reviewers.  But your case is even more interesting, because this field is inside `ContainerStatus`, which itself is a list, and so it defaults to `[]` (empty).

So even if you did (2), you don't have any `ContainerStaus` objects to initialize.  But the same is true with (4)!!

TL;DR: `Pod.Status.ContainerStatuses[]` is an empty list until someone (kubelet) populates it.  Given that someone has to explicitly populate that list, when would this default ever kick in?  Are you trying to support a new field with old clients who don't know they are supposed to set it?

I think you can still do (2) or (4), just not a CREATE time.  Given that it would be a breaking change to client-go, I'd prefer (2) if you really need to do this.  See pkg/apis/core/v1/defaults.go

**Comment by SergeyKanzhelev** [comment]
@matthyx can you please take this item?

We can keep the `null` as-is for cases when kubelet hasn't touched the field yet. What is important is to initialize `Started` as `false` and transition to `true` for `initContainers`. Ideally for Ephemeral as well.

Making this ^^^ will clean up the field semantic and also help implementing the sidecar containers as a preparation step.

**Comment by matthyx** [comment]
Sure
/assign

**Comment by matthyx** [comment]
> * The fact that we have `StartedAt` and `Started` kind of makes me grumpy, but no fixing it now.

This is because `StartedAt` refers to when the container has been created by the runtime (conversion from https://github.com/kubernetes/kubernetes/blob/98406e2a3e88bd6a2affe5674b7a577aa0926d48/pkg/kubelet/container/runtime.go#L333).
This is different from `prober_manager` having verified this state with the `startupProbe`.

**Comment by matthyx** [comment]
I think the best place to "initialize" the field would be somewhere inside:
https://github.com/kubernetes/kubernetes/blob/98406e2a3e88bd6a2affe5674b7a577aa0926d48/pkg/kubelet/kubelet_pods.go#L1664
I will provide a PR soon to spark the discussion.

**Comment by gjkim42** [comment]
> ## Proposal
> * Change `Started` type from `*bool` to `bool`.
> * This changes behavior for Init Containers and Pods before they received by the kubelet - value will be `false` instead of `null`
> * Clearly document the difference between `ContainerStatus.Started` and `ContainerStatus.State.StartedAt` fields.
> * For sidecar container KEP use the `Started` field as a signal that sidecar container started and ready to proceed with other containers initialization.

Two questions:

1. When an init container (not a sidecar container) exits with 0, what is the proper value for the `Started` field?
2. When a sidecar container exited and not yet re-starts, what is the proper value for the `Started` field?

**Comment by matthyx** [comment]
That would make sense, but requires updating again on exit... what's the behavior for `Ready`?

**Comment by gjkim42** [comment]
> what's the behavior for `Ready`?

Currently,

For regular containers:
- Ready is true if the container is in a running state and if its readinessProbe succeeds if exists.

For init containers:
- Ready is true if the container has exited with 0. (This looks exceptional.)

---

In my opinion, with sidecar containers:

For regular containers and **_sidecar containers_**:
- Ready if the container is in a running state and if its readinessProbe succeeds if exists.

For init containers (not a sidecar):
- Ready if the container has exited with 0. (This looks exceptional.)

**Comment by liggitt** [comment]
/remove-label api-review

sounds like the API review question got resolved... can re-add the label if more input is needed, or add the label to the implementing PR

**Comment by SergeyKanzhelev** [comment]
/remove-lifecycle stale

This may be double interesting for RestartPod KEP: https://github.com/kubernetes/enhancements/issues/5532
